### DISCUSSION TITLE
linear stability calculator?

### DISCUSSION THREAD

**Discussion by francispoulin** [CliMA/Oceananigans.jl#3000] [Computational science!]
Inspired by @glwagner's script from this morning with a beautiful movie of Rayeigh-Taylor instability, I thought it would be great to set up code to solve the linear stability problem for the example (and for others).  I'm not sure if others are interested so I will work to do this sometime, and this issue is a way for me not to forget about it.

**Comment by glwagner** [comment]
@navidcy put together an example that uses the "power method" for this purpose for Kelvin-Helmholz instability:

https://clima.github.io/OceananigansDocumentation/stable/generated/kelvin_helmholtz_instability/

@kburns thinks we can generalize this method to find all eigenvalues / eigenfunctions with something called "Arnoldi iteration".

Not sure this is what you had in mind, but its a start perhaps. I'm not sure how hard it would be to design a general abstraction for this purpose. If its not too much code I think it could be appropriate to go in the src. Another option is to create an independent package that uses Oceananigans (with the possibility of using other models too) for this purpose.

**Reply by JL-Cao** [reply]
hi, it's agreat example for KH instability. Well, however, when I  put the code on oceananigans KH instability on GPU, here comes so many bug. Could you please make a GPU version? After all, Julia language or Oceananigans.jl is more powerful on GPU. 
hope for your answer.....
so many thanks....

**Reply by glwagner** [reply]
It shouldn't be too difficult to make this work on GPU. I recommend eliminating the plotting first. Then wherever there is an indexing operation like `energy[1, 1, 1]`, you should use `CUDA.@allowscalar energy[1, 1, 1]`, because scalar operations are prohibited on GPU. Keep solving the bugs until the script runs.

When you're done it could be nice to share your script, either here or via a new github repo, so that other people can benefit from it.

**Reply by JL-Cao** [reply]
well, this is not the one I came across. I can run KH intability case sucessfully, well partially sucessful I'll say. In the beginning, it seems normal for me. However, after a complete model running time, the model automatically started a second run, which I never came across before. 
Maybe you guys can offer some advice?




result is as follows: 

┌ Info: Initializing simulation...
└ @ Oceananigans.Simulations /share/home/jxc301/.julia/packages/Oceananigans/soM9P/src/Simulations/run.jl:173

[00.00%] i: 0, t: 0 seconds, wall time: 22.531 minutes, max(u): (4.993e-02, 6.075e-04, 1.453e-03) m/s, next Δt: 22 seconds

┌ Info:     ... simulation initialization complete (21.629 seconds)
└ @ Oceananigans.Simulations /share/home/jxc301/.julia/packages/Oceananigans/soM9P/src/Simulations/run.jl:212
┌ Info: Executing initial time step...
└ @ Oceananigans.Simulations /share/home/jxc301/.julia/packages/Oceananigans/soM9P/src/Simulations/run.jl:114
┌ Info:     ... initial time step complete (1.250 minutes).
└ @ Oceananigans.Simulations /share/home/jxc301/.julia/packages/Oceananigans/soM9P/src/Simulations/run.jl:123

[01.54%] i: 24, t: 9.240 minutes, wall time: 2.208 minutes, max(u): (4.979e-02, 6.767e-04, 1.196e-03) m/s, next Δt: 26.620 seconds
[03.40%] i: 48, t: 20.420 minutes, wall time: 59.313 seconds, max(u): (4.969e-02, 7.262e-04, 7.650e-04) m/s, next Δt: 32.210 seconds
[05.66%] i: 72, t: 33.949 minutes, wall time: 59.356 seconds, max(u): (4.970e-02, 1.381e-03, 1.186e-03) m/s, next Δt: 38.974 seconds
[08.39%] i: 96, t: 50.318 minutes, wall time: 59.775 seconds, max(u): (4.968e-02, 3.222e-03, 2.590e-03) m/s, next Δt: 47.159 seconds
[11.69%] i: 120, t: 1.169 hours, wall time: 1.014 minutes, max(u): (4.961e-02, 5.972e-03, 4.371e-03) m/s, next Δt: 57.062 seconds
[15.59%] i: 144, t: 1.559 hours, wall time: 1.000 minutes, max(u): (4.952e-02, 1.021e-02, 8.107e-03) m/s, next Δt: 1 minute
[19.59%] i: 168, t: 1.959 hours, wall time: 59.935 seconds, max(u): (4.940e-02, 1.271e-02, 1.372e-02) m/s, next Δt: 1 minute
[23.59%] i: 192, t: 2.359 hours, wall time: 59.997 seconds, max(u): (4.940e-02, 1.558e-02, 1.619e-02) m/s, next Δt: 1 minute
[27.59%] i: 216, t: 2.759 hours, wall time: 1.001 minutes, max(u): (4.924e-02, 1.642e-02, 1.669e-02) m/s, next Δt: 1 minute
[31.59%] i: 240, t: 3.159 hours, wall time: 59.937 seconds, max(u): (4.919e-02, 1.708e-02, 1.734e-02) m/s, next Δt: 1 minute
[35.59%] i: 264, t: 3.559 hours, wall time: 59.595 seconds, max(u): (4.916e-02, 1.743e-02, 1.778e-02) m/s, next Δt: 1 minute
[39.59%] i: 288, t: 3.959 hours, wall time: 1.008 minutes, max(u): (4.917e-02, 1.633e-02, 1.765e-02) m/s, next Δt: 1 minute
[43.59%] i: 312, t: 4.359 hours, wall time: 1.000 minutes, max(u): (4.922e-02, 1.697e-02, 1.752e-02) m/s, next Δt: 1 minute
[47.59%] i: 336, t: 4.759 hours, wall time: 59.970 seconds, max(u): (4.919e-02, 1.678e-02, 1.811e-02) m/s, next Δt: 1 minute
[51.59%] i: 360, t: 5.159 hours, wall time: 59.863 seconds, max(u): (4.920e-02, 1.669e-02, 1.758e-02) m/s, next Δt: 1 minute
[55.59%] i: 384, t: 5.559 hours, wall time: 59.985 seconds, max(u): (4.916e-02, 1.654e-02, 1.724e-02) m/s, next Δt: 1 minute
[59.59%] i: 408, t: 5.959 hours, wall time: 1.009 minutes, max(u): (4.917e-02, 1.712e-02, 1.718e-02) m/s, next Δt: 1 minute
[63.59%] i: 432, t: 6.359 hours, wall time: 1.000 minutes, max(u): (4.924e-02, 1.677e-02, 1.692e-02) m/s, next Δt: 1 minute
[67.59%] i: 456, t: 6.759 hours, wall time: 1.008 minutes, max(u): (4.933e-02, 1.617e-02, 1.687e-02) m/s, next Δt: 1 minute
[71.59%] i: 480, t: 7.159 hours, wall time: 59.909 seconds, max(u): (4.931e-02, 1.612e-02, 1.697e-02) m/s, next Δt: 1 minute
[75.59%] i: 504, t: 7.559 hours, wall time: 59.965 seconds, max(u): (4.935e-02, 1.690e-02, 1.631e-02) m/s, next Δt: 1 minute
[79.59%] i: 528, t: 7.959 hours, wall time: 59.810 seconds, max(u): (4.944e-02, 1.645e-02, 1.646e-02) m/s, next Δt: 1 minute
[83.59%] i: 552, t: 8.359 hours, wall time: 59.807 seconds, max(u): (4.948e-02, 1.625e-02, 1.645e-02) m/s, next Δt: 1 minute
[87.59%] i: 576, t: 8.759 hours, wall time: 59.874 seconds, max(u): (4.948e-02, 1.611e-02, 1.615e-02) m/s, next Δt: 1 minute
[91.59%] i: 600, t: 9.159 hours, wall time: 59.776 seconds, max(u): (4.945e-02, 1.731e-02, 1.587e-02) m/s, next Δt: 1 minute
[95.59%] i: 624, t: 9.559 hours, wall time: 1.007 minutes, max(u): (4.949e-02, 1.546e-02, 1.660e-02) m/s, next Δt: 1 minute
[99.59%] i: 648, t: 9.959 hours, wall time: 59.955 seconds, max(u): (4.949e-02, 1.560e-02, 1.601e-02) m/s, next Δt: 1 minute


┌ Info: Simulation is stopping after running for 28.649 minutes.
└ @ Oceananigans.Simulations /share/home/jxc301/.julia/packages/Oceananigans/soM9P/src/Simulations/simulation.jl:180
┌ Info: Simulation time 10 hours equals or exceeds stop time 10 hours.
└ @ Oceananigans.Simulations /share/home/jxc301/.julia/packages/Oceananigans/soM9P/src/Simulations/simulation.jl:181
┌ Info: Initializing simulation...
└ @ Oceananigans.Simulations /share/home/jxc301/.julia/packages/Oceananigans/soM9P/src/Simulations/run.jl:173

[00.00%] i: 0, t: 0 seconds, wall time: 16.205 seconds, max(u): (   Inf,    Inf,    Inf) m/s, next Δt: 30 seconds

┌ Info:     ... simulation initialization complete (494.219 ms)
└ @ Oceananigans.Simulations /share/home/jxc301/.julia/packages/Oceananigans/soM9P/src/Simulations/run.jl:212
┌ Info: Executing initial time step...
└ @ Oceananigans.Simulations /share/home/jxc301/.julia/packages/Oceananigans/soM9P/src/Simulations/run.jl:114
┌ Info:     ... initial time step complete (2.811 seconds).
└ @ Oceananigans.Simulations /share/home/jxc301/.julia/packages/Oceananigans/soM9P/src/Simulations/run.jl:123

[02.10%] i: 24, t: 12.600 minutes, wall time: 1.284 minutes, max(u): (   NaN,    NaN, 0.000e+00) m/s, next Δt: 36.300 seconds
[04.64%] i: 48, t: 27.846 minutes, wall time: 1.267 minutes, max(u): (   NaN,    NaN, 0.000e+00) m/s, next Δt: 43.923 seconds
[07.72%] i: 72, t: 46.294 minutes, wall time: 1.275 minutes, max(u): (   NaN,    NaN, 0.000e+00) m/s, next Δt: 53.147 seconds
[11.44%] i: 96, t: 1.144 hours, wall time: 1.270 minutes, max(u): (   NaN,    NaN, 0.000e+00) m/s, next Δt: 1 minute

┌ Info: time = 4356.919716000007, iteration = 100: NaN found in field u. Stopping simulation.
└ @ Oceananigans.Simulations /share/home/jxc301/.julia/packages/Oceananigans/soM9P/src/Simulations/nan_checker.jl:47

2-element Vector{Any}:
 NaN
 NaN

**Reply by JL-Cao** [reply]
and as last, it will always export NaN data, which will cause model crash.

**Reply by glwagner** [reply]
I don't think that output is from the Kelvin-Helmholtz instability example. I'm not sure what code you're running

**Reply by JL-Cao** [reply]
well, this is not KH instability case exactly. This model is setup by myself to caculate some specific question about symmetric instability. Here, I just want to borrow functions used to caculate growth rate from KH case.

**Reply by JL-Cao** [reply]
`"""
    grow_instability!(simulation, energy)

Grow an instability by running `simulation`.

Estimates the growth rate ``σ`` of the instability
using the fractional change in volume-mean kinetic energy,
over the course of the `simulation`

``
energy(t₀ + Δτ) / energy(t₀) ≈ exp(2 σ Δτ)
``

where ``t₀`` is the starting time of the simulation and ``t₀ + Δτ``
the ending time of the simulation. We thus find that the growth rate
is measured by

``
σ = log(energy(t₀ + Δτ) / energy(t₀)) / (2 Δτ) .
``
"""
function grow_instability!(simulation, energy)
    # Initialize
    # σ = []
    simulation.model.clock.iteration = 0
    t₀ = simulation.model.clock.time = 0
    compute!(energy)
    energy₀ = energy[1, 1, 1]

    # Grow
    run!(simulation)  #,pickup = "./model_checkpoint_iteration8451.jld2"

    # Analyze
    compute!(energy)
    energy₁ = energy[1, 1, 1]
    Δτ = simulation.model.clock.time - t₀

    # ½(u² + v²) ~ exp(2 σ Δτ)
    σ = growth_rate = log(energy₁ / energy₀) / 2Δτ
    # push!(σ, grow_instability!(simulation, energy))
    return growth_rate
end
nothing # hide`



this is the key function caculating growth_rate. Unfortunately, I came across some bug. So I want to export velocity field first, and then use the formula to caculate directly. Could you please offer some advice?
Thanks~

**Reply by glwagner** [reply]
The velocity field components are accessible from the simulation:

```julia
u, v, w = simulation.model.velocities
```

is that what you're looking for? (I'm not sure what you mean by "export").

**Comment by francispoulin** [comment]
Yes, that is pretty much what I had in mind but for a different physical problem.  

There are a variety of ways one can do this.  Below are a few that come to mind.

1. Power method: worked well for Kelvin-Helmholtz instability and should work well for other problems.  Two nice features are that it's pretty simple, and it's already coded up for one example so should be easy to adapt.  One not so nice feature is that it only tells us the most unstable mode (spatial structure, growth rates and phase speed) but it doesn't tell us about any other unstable modes.

2. Arnoldi Method: A quick google search came across this [library](https://haampie.github.io/ArnoldiMethod.jl/stable/).  I have not tried it but could try adapting it to the above mentioned KH problem. 

3.  Standard eigenvals: If we assume periodicity in two directions it boils down to a 1D eigenvalue problem that should be easy to solve using eigenvals.  There should be a direct solver that forms a full matrix and gives you all the eigenfunctions but the indirect uses Arnoldi and gives you some.  

As for where they appera I don't kown but at some point, after I get a ShallowWaterModel off the ground, I will play around with different solvers and see what looks promising.  This won't happen this week I'm pretty sure.

**Comment by kburns** [comment]
The issue with the power method is that its convergence depends on the _difference_ between the two largest eigenvalues, and this can be really slow.  Arnoldi iteration does a lot better, and gives you multiple eigenmodes.  I think it should also be easy to add since most implementations (including ArnoldiMethod.jl, it looks like) are matrix-free.  Directly calculating the full spectrum might be harder -- Dedalus can do this because it forms sparse matrices for the global linear operator, but I don't think oceananigans does this?

**Comment by francispoulin** [comment]
Thanks @kburns for the suggestion.  I will look at Dedalus to see what they do and learn from them.  I don't think we would want to find the full spectrum unless the problem is one-dimensional, but I guess it depends on what we want it to do.  Lots of potential here I'm sure.

**Comment by kburns** [comment]
@glwagner or @ali-ramadhan would know better -- I don't think oceananigans forms a global LHS matrix for timestepping, but it might be feasible to build it just for eigenvalue problems.  But on the other hand, the oceananigans forward evaluation is so fast that I think Arnoldi would probably really rip if you can do it all on the gpu!

**Comment by glwagner** [comment]
There might be fancy ways to build the matrix using existing abstractions, but I'm not sure due to the pressure solver + fractional step stuff. With immersed boundaries I think we are out of luck for getting a matrix too, since that again makes non-trivial modifications to the discretization and time-stepping algorithm. A matrix free method that uses the user API might be easier to maintain as well. Arnoldi sounds like a pretty sweet idea...

**Comment by glwagner** [comment]
I'm going to convert this to a discussion. The example linked above illustrates how to do linear stability calculations with Oceananigans. Any other tips people have to add are welcome!

**Comment by JL-Cao** [comment]
My field of research is the study of instability mechanisms, and I am a huge fan of Oceananigans.jl. This really helps me a lot, especially KH intability you put on Oceananigans.example. I thank you for that. However, I do have some issues need handling.
(1) when I put the code on oceananigans KH instability on GPU, here comes so many bug. Could you please make a GPU version for KH instability? especially, the functions used to caculate growth-rate and rescale.
(2) it would be wonderful if you guys can put on some ohter type of instability. Well, I think the key is how you analyze the instability parameters like energy and so on.
hope for your answer, please.

**Reply by navidcy** [reply]
Hi @caojialin000,

1) Thanks for pointing out that the KH example does not work on GPU. It'd be great if you can help us fix it. There are a few tips at https://clima.github.io/OceananigansDocumentation/stable/simulation_tips/ that could be helpful. Usually it's a few minor changes that are needed, e.g. some constant appended with `const`. We'll very much appreciate a PR if you figure it out!

2) Unfortunately, we don't have any plans at the moment for putting more stability problems together. But we would very much welcome a contribution by you if you'd like! If you have any idea of a stability calculation then it'd be nice if you first put it forward here so we can all have a discussion before you go ahead putting work to it towards a PR.

**Comment by JL-Cao** [comment]
thanks. Oceananigans is a great platform for us all. I will do my best to make some contribution. 
        
  
    ---- Replied Message ----
  
  
    
      
        
         From 
        
        
            Navid C. ***@***.***>
            
        
      
      
        
         Date 
        
        
        9/10/2023 16:27
        
      
      
        
         To 
        
        
         
          
            ***@***.***>
            
          
        
      
      
        
         Cc 
        
        
          
            jialin ***@***.***>
            ,
          
          
            ***@***.***>
            
          
        
      
      
        
         Subject 
        
        
              Re: [CliMA/Oceananigans.jl] linear stability calculator? (Discussion #3000)
        
      
    
  

Hi @caojialin000,


Thanks for pointing out that the KH example does not work on GPU. It'd be great if you can help us fix it. There are a few times at https://clima.github.io/OceananigansDocumentation/stable/simulation_tips/ that could be helpful. Usually it's a few minor changes that are needed, e.g. some constant appended with const. We'll very much appreciate a PR if you figure it out!


Unfortunately, we don't have any plans at the moment for putting more stability problems together. But we would very much welcome a contribution by you if you'd like! If you have any idea of a stability calculation then it'd be nice if you first put it forward here so we can all have a discussion before you go ahead putting work to it towards a PR.

**Reply by francispoulin** [reply]
I will add that the Shallow Water model example looks at a barotropic instability, in case that is of interest to you.

**Comment by JL-Cao** [comment]
thanks~look forward to your new work, it really helps me a lot. 
        
  
    ---- Replied Message ----
  
  
    
      
        
         From 
        
        
            Francis J. ***@***.***>
            
        
      
      
        
         Date 
        
        
        9/11/2023 01:16
        
      
      
        
         To 
        
        
         
          
            ***@***.***>
            
          
        
      
      
        
         Cc 
        
        
          
            jialin ***@***.***>
            ,
          
          
            ***@***.***>
            
          
        
      
      
        
         Subject 
        
        
              Re: [CliMA/Oceananigans.jl] linear stability calculator? (Discussion #3000)
        
      
    
  

I will add that the Shallow Water model example looks at a barotropic instability, in case that is of interest to you.
